### DISCUSSION TITLE
How to use Image Magick to remove logo from image

### DISCUSSION THREAD

**Discussion by Peter455a** [ImageMagick/ImageMagick#5965] [Digital Image Processing] [answered]
I have a couple of images I need to remove logos from. I have the copy right to those images and am just doing this for learning purposes. Does anyone have any code or an approach I can use to detect and remove a logo from an image with Image Magick?

**Comment by fmw42** [comment]
What is your Imagemagick version and platform/OS?  Do you have the logo images that were used? Can you provide an example image and the logo?

**Reply by Peter455a** [reply]
Image Magick version 7.1.0-57
Os or platform is Windows 10
attached is a copy of the image 
![original](https://user-images.githubusercontent.com/58797746/212158512-b15e17fd-294d-46f1-a130-232e65f3a775.jpg)

**Comment by fmw42** [comment, marked as answer]
Do you have the logo image by itself?  If not, at least with this image, get the color of the background and just draw a filled rectangle over your logo with that color as the fill color.

`magick shoe.jpg -fill rgb(212,212,212) -draw "rectangle 64,102 %[fx:64+197],%[fx:102+194]" -alpha off show_new.jpg`

Alternately you can create a rectangle of the background color as an image and composite it over the area of the logo.

`magick shoe.jpg ( -size 197x194 xc:rgb(212,212,212) ) -geometry +64+102 -compose over -composite show_new.jpg`

![show_new](https://user-images.githubusercontent.com/13458218/212203780-f6a85579-38bf-442a-bdf7-5a3d24400121.jpg)

**Reply by Peter455a** [reply]
Can you convert the code you wrote to C#? I would like to implement the same in C#.

**Reply by fmw42** [reply]
Sorry, I do not code in C#.  Perhaps one of the developers can help you

**Reply by Peter455a** [reply]
you were running the program from the command line? Can you at least share the psuedocode so that I can try and convert to .NET code.

**Reply by Peter455a** [reply]
chat GPT generated the C# version of that code, Thank You so much for your help

**Reply by fmw42** [reply]
I posted the code and explained what I was doing. Pretty straightforward.

**Reply by Peter455a** [reply]
is your code in python?

**Reply by Peter455a** [reply]
Can you write code on how I can get the background color of a sample image and draw the background color over the logo in the image like you did above

**Reply by fmw42** [reply]
Why do you ask about Python?  This is an Imagemagick forum and I wrote code in Imagemagick command line above to do what you want if you know where the logo is and what the background color is.

If you need more, then I have more questions about what information you might have. Will the logo always be in the same location and the same size?  If the logo is not the same size and location, will you know the size of the logo?  Will you have the logo as a separate image before hand? If it has to be fully automated, then how do I distinguish the logo from the object in the image, if you have different objects in other images.

**Reply by Peter455a** [reply]
yeah most probably the logo might be in the top left of the image but in different sizes. How do install Image Magick and and then add its bin to path in the environment variables so that i can remove logo from it

**Reply by fmw42** [reply]
What platform/OS are you using?

**Reply by Peter455a** [reply]
I am using Windows 10

**Reply by Peter455a** [reply]
Is there a way to automatically detect the logo rectangle with Image Magick?

**Reply by fmw42** [reply]
I am not sure. If the background is constant and both the object and logo can be thresholded and isolated, then connected components processing can find the two largest objects - the largest is your object and the other is the logo. Or whichever one is not in the center might be the logo. Then get the bounding box of the logo and draw background color over it. See https://imagemagick.org/script/connected-components.php

**Reply by fmw42** [reply]
Sorry, I do not use Windows, so cannot advise about its installation. But see https://imagemagick.org/script/download.php#windows

**Reply by Peter455a** [reply]
Thanks for your input, you have no idea how much your help is valued.

**Reply by Peter455a** [reply]
what about the text? C image magick draw a rectangle over detected text?

**Reply by fmw42** [reply]
I do not understand the question.  Is the text in place of your logo?  Can you provide and example image?

**Reply by Peter455a** [reply]
yes the text is on the image instead of the logo. I am building a solution that checks text first and then checks logo secondly. Firstly I need to overlay all text rectangles with the background color of the image and then use the connected component technique to detect object and logo and then remove the logo based on this.  let me supply sample image

**Reply by Peter455a** [reply]
Most images that contain the text text not have a logo and below are examples

**Reply by Peter455a** [reply]
![2](https://user-images.githubusercontent.com/58797746/213517045-7b23a633-a65d-4899-b7ea-c32cad86b958.png)
![16](https://user-images.githubusercontent.com/58797746/213517098-719c06e7-8fec-4ac1-b4dc-262c0dabc2d2.jpg)

**Reply by fmw42** [reply]
Unless the text can be distinguished from the background and other objects, this would not likely be possible for Imagemagick. You should use tesseract to read the text. If the text can be be separated (thresholded) from the rest of the image, then it would be possible.

It could be done in your first image, but not easily in your second image since you have other white text in the background of the objects. Unless you know that the text in question is always in the top right corner.

**Reply by Peter455a** [reply]
Okay can you try to remove the text in the first image with Image Magick

**Comment by dlemstra** [comment]
Locking this conversation since I am seeing to many different brands in the example images.

**Reply by fmw42** [reply]
Why is that an issue?

**Comment by fmw42** [comment]
Here is one way to do that for your first example image with the logo in Unix syntax:

Input:

![shoe](https://user-images.githubusercontent.com/13458218/213538129-d4e1c8f0-fbce-466e-952e-6899806a4c08.jpg)

```
# change end of file separator so that the list is maintained in separate lines
OLDIFS=$IFS
IFS=$'\n'

# threshold the image
# use morphology to clean up small regions so not too many regions
# find the bounding boxes, filter on area, and save only the black ones
# change the x in the bounding box to + for easier extraction of w,h,x,y
bboxArr=(`magick shoe.jpg \
-color-threshold "rgb(180,180,180)-rgb(230,230,230)" \
-morphology close disk:5 \
-morphology open disk:5 \
-type bilevel \
-define connected-components:mean-color=true \
-define connected-components:area-threshold=100 \
-define connected-components:verbose=true \
-connected-components 8 tmp.png |\
grep "gray(0)" | awk '{print $2}' | tr "x" "+"`)
num=${#bboxArr[*]}
echo $num

# get the background color excluding the black regions, since IM will ignore the transparent pixels
bgcolor=`magick shoe.jpg tmp.png \
-alpha off -compose copy_opacity -composite \
-scale 1x1! -alpha off -format "%[pixel:u.p]" info:`
echo $bgcolor

# select the second largest region (since only two large regions that are not background
# note: could have selected on the one that is not in the center, but did not do that
# find the top left and bottom right x,y pixels of the bounding boxes for use in -draw
# draw the background color in a rectangle for the bounding box region
bbox=${bboxArr[1]}
echo "$bbox"
x1=`echo "$bbox" | cut -d+ -f3`
y1=`echo "$bbox" | cut -d+ -f4`
ww=`echo "$bbox" | cut -d+ -f1`
hh=`echo "$bbox" | cut -d+ -f2`
x2=$((x1+ww-1))
y2=$((y1+hh-1))
echo "$x1, $y1, $w, $h, $x2, $y2"
echo "$x1,$y1"
echo "$x2,$y2"
magick shoe.jpg -fill "$bgcolor" -draw "rectangle $x1,$y1 $x2,$y2" shoe_no_logo.jpg

# reset file separator
IFS=$OLDIFS


```

Result:

![shoe_no_logo](https://user-images.githubusercontent.com/13458218/213539841-c1bca4c7-bc4a-44c0-b6f8-01f52c61d14b.jpg)
